packages/orchestration/src/utils/zoe-tools.js

```diff
@@ -63,21 +80,102 @@ export const makeZoeTools = (zone, { zcf, vowTools }) => {
       // );
 
       // Now all the `give` are accessible, so we can move them to the localAccount`
+      const payments = await Promise.all(
+        keys(give).map(kw => E(userSeat).getPayout(kw)),
+      );
+      const settleDeposits = await when(
+        allVowsSettled(payments.map(pmt => localAccount.deposit(pmt))),
+      );
+      // if any of the deposits to LCA failed, unwind all the allocations
+      if (settleDeposits.find(x => x.status === 'rejected')) {
+        const amounts = values(give);
+        const errors = [];
+        // withdraw the successfully deposited payments
+        const paymentsOrWithdrawVs = settleDeposits.map((x, i) => {
+          if (x.status === 'rejected') {
+            errors.push(x.reason);
+            return payments[i];
+          }
+          return localAccount.withdraw(amounts[i]);
+        });
 
-      const depositVs = Object.entries(give).map(async ([kw, _amount]) => {
-        const pmt = await E(userSeat).getPayout(kw);
-        // TODO arrange recovery on upgrade of pmt?
-        return localAccount.deposit(pmt);
-      });
-      await vowTools.when(vowTools.allVows(depositVs));
+        // return all payments to the srcSeat
+        const paymentsToReturn = await when(allVows(paymentsOrWithdrawVs));
+        const paymentsKwr = harden(
+          keys(give).reduce(
+            (kwr, kw, i) => assign(kwr, { [kw]: paymentsToReturn[i] }),
+            {},
+          ),
+        );
+        const depositResponse = await depositToSeat(
+          zcf,
+          srcSeat,
+          give,
+          paymentsKwr,
+        );
+        console.debug(depositResponse);
+        throw Fail`One or more deposits to LCA failed. ${q(errors)}`;
```

Comment:
because `Fail` throws, I think we're developing some confusion by using `throw Fail`. WDYT of only using Fail for the `check || Fail` assertion pattern? 

```suggestion
        throw makeError`One or more deposits to LCA failed. ${q(errors)}`;
```

Reply:
I am onboard! This makes things clearer.

Reply:
Do you have a preference for `makeError('One or more deposits to LCA failed. ${q(errors)}')` vs `makeError('One or more deposits to LCA failed', undefined, { errors }`?

I've implemented the latter since that's what the interface was guiding towards. A draw back is that the nested error(s) only appears in logs.

Reply:
I usually defer to @Chris-Hibbert on error elision

Reply:
I think this is fine. I wouldn't expect the `errors` to reveal any sensitive details, and we believe we've returned the funds to the originator. 

How does the end user become aware of the problem? Or if this is a tool for implementors to use, how can they detect the problem?

Reply:
I think we'd want users and implementers to be aware of the error(s). Ultimately it’s implementers since this is a library function, but I think we want sensible defaults. Some of this is still an open design question for #9066, but we anticipate some level of error reporting in vstorage (at least via wallet offer results). 

If `opts.errors` is just for redaction, I would lean towards string concatenation and will refactor this.

Reply:
I'm beginning to think @turadg misdirected this question to me, and meant @kriskowal. I don't know much about `makeError`--that's deep inside **endo**.

> we anticipate some level of error reporting in vstorage
that sounds very helpful.

Reply:
I meant you @Chris-Hibbert because I recall your attention to redaction in Inter Protocol

Reply:
I don't see a problem here.

Reply:
I think that `throw makeError` is more theoretically sound, but `throw Fail` has the same effect (the `throw` is unreachable because `Fail` throws, that subsequent code is unreachable), and provides a more obvious, local indication that the following code is unreachable.

Reply:
`throw Fail` has the same effect here but not always. I think we should avoid using `throw Fail`, 
- https://github.com/Agoric/agoric-sdk/issues/10118